algo/numeric/simplex.h

```diff
@@ -0,0 +1,195 @@
+//
+// Where for all i: bs[i] >= 0, xs[i] >= 0.
+//
+// Inspired by little-known and hard-to-find, but very nice paper by
+// D.E.K.: https://www-cs-faculty.stanford.edu/~knuth/programs/lp.w
+class Simplex {
+public:
+  static inline const double kEps = 0.0000000001;
+
+  template <typename M>
+  explicit Simplex(const M& as, const std::vector<double>& bs, const std::vector<double>& fs, bool verbose = false)
+      : m_as(as.Height() + 1, as.Width() + as.Height() + 1)
+      , m_trial(m_as.Width())
+      , m_brow(m_as.Width())
+      , m_bcol(m_as.Height())
+      , m_verbose(verbose) {
+    assert(as.Height() == bs.size());
+    assert(as.Width() == fs.size());
+
+    for (size_t i = 0; i < fs.size(); ++i)
+      m_as(0, as.Height() + 1 + i) = -fs[i];
+
+    for (size_t i = 0; i < as.Height(); ++i) {
+      assert(bs[i] >= 0);
+
+      m_as(i + 1, 0) = bs[i];
+      m_as(i + 1, i + 1) = 1;
+      for (size_t j = 0; j < as.Width(); ++j)
+        m_as(i + 1, as.Height() + 1 + j) = as(i, j);
+      m_brow[i + 1] = i + 1;
+      m_bcol[i + 1] = i + 1;
+    }
+
+    if (verbose) {
+      std::cerr << "Initial table:" << std::endl;
+      Display();
+    }
+  }
+
+  // In case of finite solution, fills |target| with target function
+  // value and |xs| with variables values. Otherwise, returns false.
+  bool Solve(double& target, std::vector<double>& xs) {
+    assert(m_as.Height() > 0);
+
+    uint64_t steps = 0;
+    while (true) {
+      ++steps;
+
+      bool changed = false;
+      for (size_t j = m_as.Width() - 1; j > 0; --j) {
+        if (m_as(0, j) < 0) {
+          const auto i = GetRowToPivot(j);
+          if (i == 0) {
+            if (m_verbose)
+              std::cerr << "The maximum is infinite, proved after " << steps << " step(s)" << std::endl;
+            return false;
+          }
+          Pivot(i, j);
+          if (m_verbose) {
+            std::cerr << "After pivoting at (" << i << ", " << j << "):" << std::endl;
+            Display();
+          }
+          changed = true;
+          break;
+        }
+      }
+
+      if (!changed)
+        break;
+    }
+
+    target = m_as(0, 0);
+    xs.assign(m_as.Width() - m_as.Height(), 0.0);
+    for (size_t i = 0; i < xs.size(); ++i) {
+      const auto row = m_brow[m_as.Height() + i];
+      if (row != 0)
+        xs[i] = m_as(row, 0);
+    }
+
+    if (m_verbose) {
+      std::cerr << "Total steps: " << steps << std::endl;
+      std::cerr << "Target function value: " << target << std::endl;
+      std::cerr << "Variables values: [";
+      for (size_t i = 0; i < xs.size(); ++i) {
+        if (i != 0)
+          std::cerr << ", ";
+        std::cerr << xs[i];
+      }
+      std::cerr << "]" << std::endl;
+    }
+
+    return true;
+  }
+
+private:
+  static double Zap(double x) { return fabs(x) < kEps ? 0.0 : x; }
+
+  // Among all rows m_as[i], where i > 0, returns smallest by m_as[i] / m_as[i][j], where m_as[i][j] > 0.
```

Comment:
clarify that you mean lexicographically smallest?

Reply:
Done